sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
     memset(&local, 0, sizeof(sock_udp_ep_t));
     local.family = AF_INET6;
     local.netif  = SOCK_ADDR_ANY_NETIF;
-    local.port   = CONFIG_GCOAP_PORT;
-
+    if (IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)) {
+        local.port = CONFIG_GCOAPS_PORT;
+    } else {
+        local.port = CONFIG_GCOAP_PORT;
+    }
     int res = sock_udp_create(&_sock_udp, &local, NULL, 0);
     if (res < 0) {
         DEBUG("gcoap: cannot create sock: %d\n", res);
         return 0;
     }
 
     event_queue_init(&_queue);
-    sock_udp_event_init(&_sock_udp, &_queue, _on_sock_evt, NULL);
-    event_loop(&_queue);
 
+    if (IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)) {
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+        sock_dtls_init();
+        if (sock_dtls_create(&_sock_dtls, &_sock_udp,
+                            CREDMAN_TAG_EMPTY,
+                            SOCK_DTLS_1_2, SOCK_DTLS_SERVER) < 0) {
+            puts("gcoap: error creating DTLS sock");
+            sock_udp_close(&_sock_udp);
+            return 0;
+        }
+        dsm_init();
+        sock_dtls_event_init(&_sock_dtls, &_queue, _on_sock_dtls_evt,
+                            NULL);
+#endif
+    } else {
+        sock_udp_event_init(&_sock_udp, &_queue, _on_sock_udp_evt, NULL);
+    }
+
+    event_loop(&_queue);
     return 0;
 }
 
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+/* Handles DTLS socket events from the event queue */
+static void _on_sock_dtls_evt(sock_dtls_t *sock, sock_async_flags_t type, void *arg) {
+    (void)arg;
+    coap_socket_t socket = { .type = COAP_SOCKET_TYPE_DTLS, .socket.dtls = sock};
+
+    if (type & SOCK_ASYNC_CONN_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &socket.ctx_dtls_session,
+                            _listen_buf, sizeof(_listen_buf),
+                            CONFIG_GCOAP_DTLS_HANDSHAKE_TIMEOUT_USEC);
+        if (res != -SOCK_DTLS_HANDSHAKE) {
+            DEBUG("gcoap: could not establish DTLS session: %d\n", res);
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
+            return;
+        }
+        dsm_state_t prev_state = dsm_store(sock, &socket.ctx_dtls_session,
+                                           SESSION_STATE_ESTABLISHED, false);
+
+        /* If session is already stored and the state was SESSION_STATE_HANDSHAKE
+        before, the handshake has been initiated internally by a gcoap client request
+        and another thread is waiting for the handshake. Send message to the
+        waiting thread to inform about established session */
+        if (prev_state == SESSION_STATE_HANDSHAKE) {
+            msg_t msg = { .type = DTLS_EVENT_CONNECTED };
+            msg_send(&msg, _auth_waiting_thread);
+        } else if (prev_state == NO_SPACE) {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions stored in tinydtls internally */
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
```

Comment:
if it shouldn't happen, maybe add some debug output / `assert(false)`

Reply:
Added debug output here. Not sure whether `assert(false)` is really necessary here.